pkg/registry/pool/memory_pool_allocator.go

```diff
@@ -0,0 +1,148 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pool
+
+import (
+	mathrand "math/rand"
+	"sync"
+
+	"time"
+)
+
+// MemoryPoolAllocator is PoolAllocator that is backed by an in-memory collection,
+// can't be used if there are multiple processes allocating from the pool (cf EtcdPoolAllocator)
+type MemoryPoolAllocator struct {
+	driver         PoolDriver
+	randomAttempts int
+
+	lock   sync.Mutex // protects 'used' and 'random'
+	used   map[string]Allocation
```

Comment:
this is really expensive - you end up storing every key twice, and you've expanded what was a bitmap into a set.  I didn't follow all of the discussion on whether the ip allocator was properly consistent, but this seems to be a pretty large departure, with not much justification (so far)